### ISSUE TITLE
Audit 1.4.3 — performance, queries & caching (29 items)

### ISSUE THREAD

**Issue by gin0115** [a8cteam51/internet-archive-wayback-machine-link-fixer#359]
29 issues, four of them Critical. The multiplier under most of the rest is the first one: the link
table has no index on any queried column, so every lookup that isn't by primary key is a full scan.
Fix that before the N+1 patterns, because they're merely wasteful against an indexed table and
pathological against a scan. Note that the index needs a migration, and migrations currently never
run on update — `S007` in group 6 blocks this.

The other clusters: four caches that don't cache (one tests the wrong thing, one caches the cheap
answer and not the expensive one), six queries with no limit, two HTTP timeouts set in seconds but
documented and valued as milliseconds, and a check history that grows forever and is inlined into
every page's HTML twice.

Nothing here is re-verified against the source — severities and locations are the audit's.
`S###` = structured audit, `T###` = exploratory pass.

- [ ] **S010** · High · `migrations/Migration_1.php:44` — the table is created with `PRIMARY KEY (id)` and nothing else. `url` and `redirect_url` are `longtext` (`:46`, `:51`), which MySQL cannot index without a prefix length, so this can't be fixed by adding an index alone. `find_by_url()` runs `WHERE url = %s OR redirect_url = %s` (`Link_Repository.php:81`) — a full scan per call, invoked per link from `find_or_create()` and from the unauthenticated REST route. Seven exploratory lenses found this independently.
- [x] **S001** · Critical · `Settings/Settings.php:167` — `get_link_checker_timeout()` returns 5000 and its docblock at `:160` says "in MS". `HTTP_Link_Checker_Client` assigns it at `:38` and passes it to `wp_safe_remote_get()` at `:90`, and `WP_Http` reads timeout in **seconds**. A hung endpoint pins a PHP worker for up to 83 minutes, reachable from the unauthenticated REST route. Four structured reviewers and seven exploratory lenses.
- [x] **T119** · Confirmed · `Wayback_Machine/HTTP_Client/HTTP_Snapshot_Client.php:173` — the same unit error on the snapshot POST: `'timeout' => apply_filters('iawmlf_create_snapshot_timeout', 1000)` — 1000 seconds, ~16.6 minutes. Runs inside Action Scheduler workers *and* synchronously in the admin for bulk actions of five or fewer links.
- [x] **S012** · High · `Wayback_Machine/HTTP_Client/HTTP_Link_Checker_Client.php:150` — `check_single()` calls `get_final_url()` at `:150`, which performs a full GET and returns only `$response['location']`, discarding the decoded body — then `:152` queries the same endpoint again. Two blocking round trips per check. `Report_Table::process_check_action()` loops it per link (`:335-336`), so a bulk check on 50 links fires 100 sequential external calls inside one admin request.
- [x] **S004** · Critical · `Link/Link.php:508` — `jsonSerialize()` emits the whole `checks` array and nothing ever prunes it: `add_check()` (`:245`) only appends, and the only `array_slice` is a read in `is_valid()` (`:351`). It's inlined into every page twice — `wp_localize_script` at `WP_Post_Controller.php:258` and an `esc_attr`'d data attribute at `:322`. Growth is externally driven because the public REST endpoint can trigger checks anonymously.
- [ ] ~~**S002** · Critical · `WP_Post/WP_Post_Table_Controller.php:87` — `get_link()` guards its cache with `array_key_last( $this->links ) === $link_id` instead of `array_key_exists()`, so only the most recently stored ID can hit while the array grows regardless (`:95`). On the standard 20-posts-per-page screen with 10 links per post that's up to 200 full-scan SELECTs for one admin page.~~
- [x] **S038** · Medium · `functions.php:514` — `iawmlf_is_archive_api_online()` tests `if ( false !== (bool) $online && false === $force )`, which reduces to `$online === true`, so a cached `false` never satisfies it. Independently, `get_transient()` returns `false` for both a stored false and a missing key. The positive answer caches for an hour and the negative answer has an effective TTL of zero — backwards, since the negative case is the expensive one.
- [x] **S064** · Medium · `Dashboard/Dashboard_Notifications.php:137` — `get_account_details()` has its `set_transient` inside `if ( is_array( $details ) )` (`:137-138`); the `catch` returns null at `:147` and the trailing `return null` at `:150` write nothing. A site with no or invalid keys therefore makes a blocking archive.org request on every dashboard load.
- [ ] ~~**S029** · Medium · `Link/Link_Repository.php:373` — `Link_Exclusion::get_instance()` returns `new self()` with no memoisation despite the name, and its constructor calls `get_bundled_link_exclusions()` and `get_link_exclusions()` (the latter doing `get_option()` + `array_map('esc_html')` + `apply_filters`). It's called inside the per-link loop. `Plugin::get_instance()` does hold a static instance, so the pattern exists — it just wasn't applied.~~
- [x] **S003** · Critical · `Ajax/Post_Search_Ajax.php:130` — `WP_Query` with `'posts_per_page' => -1`, no `no_found_rows`, and a `posts_where` filter adding `post_title LIKE '%term%' OR post_name LIKE '%term%'` (`:166-170`). `MIN_CHARS` is 2 (`:39`), so typing `th` returns every match fully hydrated with `SQL_CALC_FOUND_ROWS` on top; the leading wildcard defeats any index. (The audit rejected one part of the original claim: `'fields' => 'ids'` wouldn't help, since `format_result_from_post()` needs the objects.)
- [ ] ~~**S013** · High · `Link/Link_Repository.php:713` — `get_all_link_meta()` runs `SELECT * FROM {postmeta} WHERE meta_key = %s` with no LIMIT and no column list, then `maybe_unserialize()`s every row (`:730-741`). One row per scanned post, so a 50k-post site materialises 50k serialised arrays in one request. Called from `Dashboard_Page.php:287` and `Report_Page.php:380`.~~
- [ ] ~~**T039** · Confirmed · `Dashboard/Dashboard_Statistics.php:420` — `compile_link_statistics()` issues eight separate `SELECT COUNT(*)` queries plus one `query_links()` against the unindexed table, every one passing `PHP_INT_MAX` as the limit. The transient expires after 2 minutes (`:353`), so a moderately trafficked wp-admin regenerates all nine every two minutes.~~
- [ ] ~~**S031** · Medium · `Dashboard/Dashboard_Statistics.php:301` — onboarding post counts run an unbounded `WP_Query` with a `NOT EXISTS` meta join and `SQL_CALC_FOUND_ROWS`. Its OR clause is also dead — it compares a serialised array of link IDs against `time()`, same defect as `S024` in group 5.~~
- [ ] ~~**S030** · Medium · `Util/Action_Scheduler_Garbage_Collection.php:79` — `'per_page' => -1` in all four cleanup methods (`:66`, `:149`, `:232`, `:315`), loading every failed action ID, then a separate `$store->fetch_action()` per ID (`:87-99` and three siblings) — an N+1 with unbounded N, all four run in one cron pass.~~
- [ ] ~~**T233** · Minor · `Processor/Post_Processor.php:55` — `process()` maps `find_or_create()` over every unique external link with no cap; `Content_Scanner::scan()` applies `array_unique()` but no limit (`:87`). Each call is a full scan plus, for new URLs, an INSERT and an Action Scheduler enqueue.~~
- [ ] ~~**T102** · Confirmed · `WP_Post/WP_Post_Controller.php:97` — link processing runs synchronously on `save_post` at priority 10 (`:61`, `:97`), so publishing a post with 100 external links does 100 unindexed full scans on the editor's request thread before the page returns.~~
- [ ] ~~**S011** · High · `Link/Link_Repository.php:356` — `get_links_for_post()` reads the meta array of link IDs then `array_map`s `find_by_id()` over it (`:356-361`) — one SELECT per link. On the front end it's called twice for the same post with no memoisation, at `WP_Post_Controller.php:231` and `:315`, so a post with 40 links costs 80 SELECTs per page view. `build_query()` already supports `$link_ids` (`:544`), so a batched path exists unused.~~
- [ ] **S032** · Medium · `Dashboard/Dashboard_Page.php:280` — `render_page()` calls `query_links()` at `:190`, which returns fully hydrated `Link` objects, then `:188` maps each through `get_link_stats()`, which calls `find_by_id()` (`:281`) to re-fetch by primary key the exact rows it just had — plus one `get_post` per listed link.
- [ ] **S105** · Low · `Dashboard/Dashboard_Page.php:176` — `get_link_statistics()` is called at `:176` and again at `:249`, each doing its own `get_transient()` read, and `get_links_per_page()` is likewise called twice.
- [ ] ~~**S027** · Medium · `Link/Link_Repository.php:603` — `count_links()` forwards `$limit` and `$page` into `build_query()`, which appends `LIMIT %d OFFSET %d` (`:607`) to a `SELECT COUNT(*)` (`:522`). An aggregate with no GROUP BY returns one row, so any OFFSET above zero returns none and the count comes back 0. Every current caller passes page 1, so it's latent. The COUNT also carries the full ORDER BY including `JSON_EXTRACT` expressions.~~
- [ ] **T071** · Confirmed · `Link/Link_Repository.php:649` — `compile_order_by()` guards `ORDER_DATE_DESC` with `CASE WHEN JSON_LENGTH(checks) = 0` before touching the JSON path (`:656-658`), but `ORDER_DATE_ASC` has no such guard and evaluates `JSON_EXTRACT(checks, CONCAT('$[', JSON_LENGTH(checks) - 1, '].date'))` inside its CASE (`:649-653`). The date filter at `:556` is likewise unguarded. `'[]'` is the normal state of a never-checked link, and MySQL 8 raises `ERROR 3143`.
- [ ] ~~**S028** · Medium · `Link/Link_Repository.php:259` — every UPDATE is followed by a redundant SELECT to rehydrate the row it just wrote.~~
- [ ] ~~**S015** · High · `assets/js/build/front_link_checker.js` — the bundle does `document.querySelectorAll('a')` and attaches an `IntersectionObserver` to every anchor on the page — nav, footer, sidebar — regardless of whether it appears in the payload. It then attaches a `MutationObserver` to `document.body` with `{childList: true, subtree: true}` whose handler does `i = [...i, ...r]`, reallocating the whole array on every mutation with no dedupe, so it grows without bound on any dynamic page. The seen-set is an array queried with `.includes()`.~~
- [x] **S014** · High · `Event/Event_Controller.php:41` — three callbacks hooked on `init`, each calling `as_has_scheduled_action()` / `as_next_scheduled_action()`, which resolve to an uncached query against `actionscheduler_actions`. Graded **Tentative** and partly overstated: two of the three return early unless their feature is enabled, so the floor is one uncached query per request, not three.
- [ ] ~~**S034** · Medium · `WP_Post/WP_Post_Controller.php:230` — `enqueue_frontend_script()` fetches post links on every front-end page including archives, using `get_the_ID()` at `wp_enqueue_scripts` time — before the loop has started, so on an archive the localised payload matches at most one post while `render_block()` emits spans for all of them.~~
- [ ] ~~**S033** · Medium · `WP_Post/WP_Post_Controller.php:287` — `render_block` runs `Settings::should_render_html_link_output()` for every block of every page.~~
- [x] **S102** · Low · `Dashboard/Settings_Page.php:49` — 21 `register_setting` calls, 4 sections and 20 fields are registered on every `admin_init`, including `admin-ajax` requests.
- [ ] ~~**S103** · Low · `Processor/Content_Scanner.php:102` — both site URLs are re-normalised once per scanned link rather than hoisted out of the loop.~~
- [ ] ~~**S104** · Low · `Settings/Settings.php:753` — `get_link_icon_css()` builds the icon list twice per call.~~

**Comment by wpvitesse-pro** [comment]
# Root cause cluster

The fundamental issue is **structural pathology from an unindexed query column**: `url` and `redirect_url` are `longtext` columns queried via equality but never indexed, forcing a full table scan for every `find_by_url()` call. Combined with N+1 anti-patterns and unbounded result sets, this transforms linear operations into quadratic behavior. The HTTP timeout bug (`5000ms` interpreted as **5000 seconds**) compounds this by holding PHP workers hostage during external calls, and the uncapped `checks` array converts every successful scan into permanent memory bloat.

---

## Fix sequence (9 discrete changes)

### 1. Index the link lookup columns
**File:** `migrations/Migration_2.php` (new)  
**Why:** `longtext` can't be indexed without a prefix. Convert to `varchar(2083)` (RFC 3986 max URL length) so MySQL can build a B-tree index.

```php
<?php
class Migration_2 {
    public function up() {
        global $wpdb;
        $table = $wpdb->prefix . 'iawmlf_links';
        
        // Modify columns to indexable type
        $wpdb->query( "ALTER TABLE `{$table}` 
            MODIFY COLUMN `url` VARCHAR(2083) NOT NULL,
            MODIFY COLUMN `redirect_url` VARCHAR(2083) DEFAULT NULL" );
        
        // Add composite index for find_by_url() query
        $wpdb->query( "ALTER TABLE `{$table}` 
            ADD INDEX `idx_url_lookup` (`url`, `redirect_url`)" );
    }
}
```

Register in your migration runner and **fix S007** first so it actually executes on update.

---

### 2. HTTP timeout units
**File:** `Settings/Settings.php:167`  
**Change:** Return value in **seconds**, not milliseconds.

```php
public function get_link_checker_timeout(): int {
    return 5; // WP_Http expects seconds
}
```

**File:** `Wayback_Machine/HTTP_Client/HTTP_Snapshot_Client.php:173`

```php
'timeout' => apply_filters( 'iawmlf_create_snapshot_timeout', 10 ) // 10s max
```

This prevents the 83-minute worker pin scenario from S001/T119.

---

### 3. Deduplicate HTTP calls
**File:** `HTTP_Link_Checker_Client.php:150-152`
